### PR TITLE
OpenVR + Vulkan support

### PR DESCRIPTION
**Pull request by phr00t**
Adds Vulkan support to OpenVR/SteamVR, so we don't have to rely on outdated DirectX 11 anymore!

## Motivation and Context

DirectX 11 is slow. VR needs fast. Vulkan is fast. They belong together.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: sources/targets/Xenko.GlobalSettings.targets [hunk 1]
```diff
@@ -4,11 +4,11 @@
   <!-- Default values -->
   <PropertyGroup>
     <!-- Defaults for CPU and GraphicsApi based on platform-->
-    <XenkoDefaultGraphicsApi Condition="'$(XenkoPlatform)' == 'Windows'">Direct3D11</XenkoDefaultGraphicsApi>
+    <XenkoDefaultGraphicsApi Condition="'$(XenkoPlatform)' == 'Windows'">Vulkan</XenkoDefaultGraphicsApi>
```
Comment (xen2) [inline sources/targets/Xenko.GlobalSettings.targets]: Please remove that from PR (you can keep it in personal branch)

File: sources/targets/Xenko.GlobalSettings.targets [hunk 2]
```diff
@@ -4,11 +4,11 @@
...
     <XenkoDefaultGraphicsApi Condition="'$(XenkoPlatform)' == 'UWP'">Direct3D11</XenkoDefaultGraphicsApi>
     <XenkoDefaultGraphicsApi Condition="'$(XenkoPlatform)' == 'Android'">OpenGLES</XenkoDefaultGraphicsApi>
     <XenkoDefaultGraphicsApi Condition="'$(XenkoPlatform)' == 'iOS'">OpenGLES</XenkoDefaultGraphicsApi>
-    <XenkoDefaultGraphicsApi Condition="'$(XenkoPlatform)' == 'Linux'">OpenGL</XenkoDefaultGraphicsApi>
+    <XenkoDefaultGraphicsApi Condition="'$(XenkoPlatform)' == 'Linux'">Vulkan</XenkoDefaultGraphicsApi>
```
Comment (xen2) [inline sources/targets/Xenko.GlobalSettings.targets]: Please remove that from PR (you can keep it in personal branch)

Comment (xen2) [reply sources/targets/Xenko.GlobalSettings.targets]: (note: totally fine to switch Linux to Vulkan by default soon, just need little bit of testing)

File: sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs [hunk 3]
```diff
@@ -1,18 +1,23 @@
 // Copyright (c) Xenko contributors (https://xenko.com) and Silicon Studio Corp. (https://www.siliconstudio.co.jp)
 // Distributed under the MIT license. See the LICENSE.md file in the project root for more information.
-#if XENKO_GRAPHICS_API_DIRECT3D11
+#if XENKO_PLATFORM_WINDOWS_DESKTOP
```
Comment (xen2) [inline sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs]: If enabling OpenVR on Windows in general, does OpenGL/OpenGLES still compiles?
Also, I think D3D11 was covering UWP too before (not sure if OpenVR was working on that platform either, if not it's OK for it to not be active anymore)

Comment (phr00t) [reply sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs]: Hrm.. I get no build errors, although I'm unsure if OpenGL/OpenGLES is trying to build...

File: sources/engine/Xenko.Engine/Rendering/Compositing/ForwardRenderer.cs [hunk 4]
```diff
@@ -172,13 +172,13 @@ protected override void InitializeCore()
...
                     vrSystem.PreferredScalings = preferredScalings;
 
-#if XENKO_GRAPHICS_API_VULKAN
-                    // no mirror support for vulkan
-                    vrSystem.RequireMirror = false;
-                    VRSettings.CopyMirror = false;
-#else
-                    vrSystem.RequireMirror =  VRSettings.CopyMirror;
-#endif
+                    if( GraphicsDevice.Platform == GraphicsPlatform.Vulkan ) {
```
Comment (xen2) [inline sources/engine/Xenko.Engine/Rendering/Compositing/ForwardRenderer.cs]: Please follow the global style (opening curly brace on new line)

Comment (xen2) [reply sources/engine/Xenko.Engine/Rendering/Compositing/ForwardRenderer.cs]: Also space after if but not inside parenthesis, i.e. `if (GraphicsDevice.Platform == GraphicsPlatform.Vulkan)`

Comment (xen2) [reply sources/engine/Xenko.Engine/Rendering/Compositing/ForwardRenderer.cs]: Never mind, fixed it directly with github online editor :)
Please try to follow those rules for other commits.

File: sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs [hunk 5]
```diff
@@ -201,6 +206,22 @@ static OpenVR()
 
         public static bool InitDone = false;
 
+#if XENKO_GRAPHICS_API_VULKAN
+        private static unsafe VRVulkanTextureData_t vkTexData;
+        public static bool InitVulkan(GameBase baseGame) {
```
Comment (xen2) [inline sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs]: curly brace style

File: sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs [hunk 6]
```diff
@@ -201,6 +206,22 @@ static OpenVR()
...
+            vkTexData = new VRVulkanTextureData_t {
```
Comment (xen2) [inline sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs]: curly brace style

File: sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs [hunk 7]
```diff
@@ -227,21 +247,37 @@ public static void Shutdown()
...
         {
-            var tex = new Texture_t
-            {
-                eType = EGraphicsAPIConvention.API_DirectX,
-                eColorSpace = EColorSpace.Auto,
-                handle = texture.NativeResource.NativePointer,
-            };
-            var bounds = new VRTextureBounds_t
-            {
+            var bounds = new VRTextureBounds_t {
```
Comment (xen2) [inline sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs]: curly brace style

File: sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs [hunk 8]
```diff
@@ -227,21 +247,37 @@ public static void Shutdown()
...
                 vMin = viewport.Y,
                 vMax = viewport.Height,
             };
-
+#if XENKO_GRAPHICS_API_VULKAN
+            vkTexData.m_nHeight = (uint)texture.Height;
+            vkTexData.m_nWidth = (uint)texture.Width;
+            vkTexData.m_nImage = (ulong)texture.NativeImage.NativeHandle;
+            vkTexData.m_nSampleCount = texture.IsMultisample ? (uint)texture.MultisampleCount : 1;
+            unsafe {
```
Comment (xen2) [inline sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs]: curly brace style

File: sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs [hunk 9]
```diff
@@ -227,21 +247,37 @@ public static void Shutdown()
...
+                fixed(VRVulkanTextureData_t* vkAddress = &vkTexData) {
```
Comment (xen2) [inline sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs]: curly brace style

**Comment by xen2** [comment]
FYI, I will take the time to review this properly right after GDC. Sorry for the delay.

File: sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs (hunk 3 above)
Comment (xen2) [reply sources/engine/Xenko.VirtualReality/OpenVR/OpenVR.cs]: Various builds are broken.
https://teamcity.xenko.com/project.html?projectId=Engine&tab=projectOverview&branch_Engine=pull%2F407%2Fmerge&guest=1

Does it work on Vulkan Windows and Linux/OSX too?
if yes, the proper define would be `XENKO_GRAPHICS_API_DIRECT3D11 || XENKO_GRAPHICS_API_VULKAN`

**Pull request closed without merging**